### PR TITLE
QAM Migration scenarios

### PR DESCRIPTION
**Pull request by dasantiago**
scenarios ready:
	* migration_zdup_offline_sle12sp1: http://tragicbox.suse.cz/tests/1410
	* migration_offline_sle12_ltss+allpatterns_fullupdate: http://tragicbox.suse.cz/tests/1409
	* migration_offline_sle12sp1: http://tragicbox.suse.cz/tests/1407
	* migration_online_sle12_ltss+allpatterns_fullupdate: http://tragicbox.suse.cz/tests/1408
	* migration_online_sle12sp1: http://tragicbox.suse.cz/tests/1411

Jobs:

### CODE REVIEW

File: tests/update/patch_before_migration.pm [hunk 1]
```diff
@@ -40,16 +40,26 @@ sub patching_sle {
...
     if (get_var('FULL_UPDATE')) {
         fully_patch_system();
         type_string "reboot\n";
         $self->wait_boot(textmode => !is_desktop_installed, ready_time => 600);
+        # Go back to the initial state, before the patching
         $self->setup_migration;
     }
 
-    de_register(version_variable => 'HDDVERSION');
+    if(get_var('FLAVOR','') eq 'Server-DVD-Updates' || get_var('FLAVOR','') =~ /David$/) {
```
Comment (dasantiago) [inline tests/update/patch_before_migration.pm]: Forgot to remove my flavour :-(

File: tests/online_migration/sle12_online_migration/post_migration.pm [hunk 2]
```diff
@@ -14,13 +14,16 @@ use base "installbasetest";
...
 use utils;
+use qam qw/add_test_repositories remove_test_repositories/;
 
 sub run {
     select_console 'root-console';
 
     # print repos to screen and serial console after online migration
     zypper_call('lr -u');
 
+    add_maintenance_repos() if(get_var('MAINT_TEST_REPO',''));
```
Comment (okurz) [inline tests/online_migration/sle12_online_migration/post_migration.pm]: why are you supplying a default when you just want to check for existance? Also, please add a space after `if`

File: tests/update/patch_before_migration.pm [hunk 3]
```diff
@@ -40,16 +40,26 @@ sub patching_sle {
...
+        # The system is registered.
+        set_var('HDD_SCC_REGISTERED',1);
```
Comment (okurz) [inline tests/update/patch_before_migration.pm]: please add a space after the comma

File: lib/qam.pm [hunk 4]
```diff
@@ -74,16 +74,43 @@ sub is_patch_needed {
 # Function that will add all test repos
 sub add_test_repositories {
     my $counter = 0;
+
+    my $oldrepo = get_var('PATCH_TEST_REPO', '');
```
Comment (coolo) [inline lib/qam.pm]: as you check $oldrepo anyway, there is no need to pass '' everywhere.

File: tests/update/patch_before_migration.pm [hunk 5]
```diff
@@ -40,16 +40,27 @@ sub patching_sle {
...
         type_string "reboot\n";
-        $self->wait_boot(textmode => !is_desktop_installed, ready_time => 600);
-        $self->setup_migration;
+        $self->wait_boot(textmode => !is_desktop_installed(), ready_time => 600);
+        # Go back to the initial state, before the patching
+        $self->setup_migration();
     }
 
-    de_register(version_variable => 'HDDVERSION');
+    if (get_var('FLAVOR', '') eq 'Server-DVD-Updates') {
```
Comment (coolo) [inline tests/update/patch_before_migration.pm]: this is check_var recoded.

But I wonder if this FLAVOR check will scale. Maintenance project managers want incidents to be tested - and they have a different flavor. So possibly we need a different variable they have in common?

File: lib/qam.pm [hunk 6]
```diff
@@ -74,16 +74,43 @@ sub is_patch_needed {
...
+    my $output   = script_output("zypper lr");
+    my @repoList = ();
+    for my $line (split(/\n/, $output)) {
+        my $repo = (split(/\|/, $line))[1];
+        if ($repo =~ /TEST_\d+/) {
+            $repo =~ s/^\s+|\s+$//g;
+            push @repoList, $repo;
+        }
+    }
+    zypper_call('rr ' . join(' ', @repoList)) if @repoList;
```
Comment (coolo) [inline lib/qam.pm]: pro tip: better parse zypper lr -e - | grep name=TEST

Comment (dasantiago) [inline lib/qam.pm]: I'm looking for the alias, but I'm not really understanding what you mean... Do you mean an one-liner?

`zypper lr $(zypper lr -e - | perl -e 'while(<>){if(/\[(repo.*)\]/){push @a,$1;}} $,=" ";print @a')`

File: lib/qam.pm (hunk 4 above)
Comment (dasantiago) [inline lib/qam.pm]: fixed

File: tests/update/patch_before_migration.pm (hunk 5 above)
Comment (dasantiago) [inline tests/update/patch_before_migration.pm]: Just changed the regexp to /-Updates$/. 

Is it better?

File: lib/qam.pm (hunk 6 above)
Comment (coolo) [inline lib/qam.pm]: hmm, the alias and the name are the same for the TEST repos afaik. And so my one liner would look like
```
zypper rr $(zypper lr -e - | grep name=TEST | cut -d= -f2)
```
But that will fail if there are no test repos

File: tests/update/patch_before_migration.pm (hunk 5 above)
Comment (coolo) [inline tests/update/patch_before_migration.pm]: That surely fixes it for SLED - which I'm not sure we want to test, but I was refering to FLAVOR=Server-DVD-Incidents 

But let's have these test cases in Updates first.

File: lib/qam.pm (hunk 6 above)
Comment (dasantiago) [inline lib/qam.pm]: This case won't fail:

`repos=($(zypper lr -e - | grep name=TEST | cut -d= -f2));  echo $repos; if [ ${#repos[@]} -ne 0 ]; then zypper rr $repos; fi`

Comment (dasantiago) [inline lib/qam.pm]: I changed the code to reflect the modifications suggested. I changed the remove_test_repos and the add_test_repos.

As soon as the test finish i'll remove the WIP

File: tests/update/patch_before_migration.pm (hunk 5 above)
Comment (dasantiago) [inline tests/update/patch_before_migration.pm]: Change done to address your concern

**Pull request merged**